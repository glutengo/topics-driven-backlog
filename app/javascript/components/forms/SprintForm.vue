<template>
    <b-form @submit.prevent="submit">
        <b-form-group label="Name" label-for="sprint-name">
            <b-form-input id="sprint-name" v-model="data.name" required></b-form-input>
        </b-form-group>

        <b-form-group label="Start date" label-for="sprint-start-date">
            <b-form-input id="sprint-start-date" :max="data.end_date" type="date" v-model="data.start_date" required></b-form-input>
        </b-form-group>

        <b-form-group label="End date" label-for="sprint-end-date">
            <b-form-input id="sprint-end-date" type="date" v-model="data.end_date" required
                          :min="$store.state.currentDate > data.start_date ? $store.state.currentDate : data.start_date"></b-form-input>
        </b-form-group>

        <b-button type="submit" variant="primary" :disabled="isInvalid">Save</b-button>
        <b-button type="button" variant="grey" @click="cancel">Cancel</b-button>
    </b-form>
</template>

<script>
    import {required} from 'vuelidate/lib/validators';
    import smallerOrEqualThan from '../../validators/smallerOrEqualThan';
    import largerOrEqualThan from '../../validators/largerOrEqualThan';
    import BaseForm from './BaseForm.vue';
    import moment from "moment";

    export default {
        name: 'sprint-form',
        extends: BaseForm,
        validations: {
            data: {
                name: {required},
                start_date: {
                    required,
                    smallerThanEnd: smallerOrEqualThan('end_date')
                },
                end_date: {
                    required,
                    largerThanStart: largerOrEqualThan('start_date'),
                    largerThanToday: largerOrEqualThan(moment().format('YYYY-MM-DD'), true),
                },
            },
        },
    }
</script>