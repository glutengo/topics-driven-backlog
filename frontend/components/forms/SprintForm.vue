<template>
    <BForm 
        @submit.prevent="submit" 
        class="sprint-form">

        <BFormGroup 
            label="Name" 
            label-for="sprint-name">

            <BFormInput 
                id="sprint-name" 
                v-model="data.name" 
                required/>
        </BFormGroup>

        <BFormGroup 
            label="Start date" 
            label-for="sprint-start-date">

            <BFormInput 
                id="sprint-start-date" 
                :max="data.end_date" 
                type="date" 
                v-model="data.start_date" 
                required/>
        </BFormGroup>

        <BFormGroup 
            label="End date" 
            label-for="sprint-end-date">

            <BFormInput 
                id="sprint-end-date" 
                type="date" 
                v-model="data.end_date" 
                required
                :min="$store.state.currentDate > data.start_date ? $store.state.currentDate : data.start_date"/>
        </BFormGroup>

        <BButton 
            type="submit" 
            variant="primary" 
            :disabled="isInvalid">Save</BButton>

        <BButton 
            type="button" 
            variant="grey" 
            @click="cancel">Cancel</BButton>
    </BForm>
</template>

<script>
import moment from 'moment';
import { required } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BButton from '@bootstrap/button/button';
import smallerOrEqualThan from '@/validators/smallerOrEqualThan';
import largerOrEqualThan from '@/validators/largerOrEqualThan';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'SprintForm',
    components: { BForm, BFormGroup, BFormInput, BButton },
    extends: BaseForm,
    validations: {
        data: {
            name: { required },
            start_date: {
                required,
                smallerThanEnd: smallerOrEqualThan('end_date'),
            },
            end_date: {
                required,
                largerThanStart: largerOrEqualThan('start_date'),
                largerThanToday: largerOrEqualThan(
                    moment().format('YYYY-MM-DD'),
                    true
                ),
            },
        },
    },
};
</script>