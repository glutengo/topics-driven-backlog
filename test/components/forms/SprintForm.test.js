import { mount, createLocalVue } from 'vue-test-utils';
import moment from 'moment';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import SprintForm from '@/components/forms/SprintForm';

// take a date far in the future so validations pass
Date.now = jest.fn(() => 7956915742000);

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('SprintForm.test.js', () => {
    let cmp, store;

    beforeEach(() => {
        store = new Vuex.Store({
            currentDate: moment().format('YYYY-MM-DD'),
        });

        cmp = mount(SprintForm, {
            localVue,
            store,
            propsData: {
                data: {
                    name: 'Test Sprint',
                    start_date: moment().format('YYYY-MM-DD'),
                    end_date: moment().add(5, 'days').format('YYYY-MM-DD'),
                },
            },
        });
    });

    it('has the expected html structure', () => {
        expect(cmp.element).toMatchSnapshot();
    });
});