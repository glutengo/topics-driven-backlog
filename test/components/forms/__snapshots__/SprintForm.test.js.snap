// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintForm.test.js has the expected html structure 1`] = `
<form
  class="sprint-form"
>
  <div
    aria-labelledby="__BVID__15___BV_label_"
    class="b-form-group form-group"
    id="__BVID__15_"
    role="group"
  >
    <label
      class="col-form-label pt-0"
      for="sprint-name"
      id="__BVID__15___BV_label_"
    >
      Name
    </label>
    <div
      class=""
    >
      <input
        aria-required="true"
        class="form-control"
        id="sprint-name"
        required="required"
        type="text"
      />
      <!---->
      <!---->
      <!---->
    </div>
  </div>
   
  <div
    aria-labelledby="__BVID__17___BV_label_"
    class="b-form-group form-group"
    id="__BVID__17_"
    role="group"
  >
    <label
      class="col-form-label pt-0"
      for="sprint-start-date"
      id="__BVID__17___BV_label_"
    >
      Start date
    </label>
    <div
      class=""
    >
      <input
        aria-required="true"
        class="form-control"
        id="sprint-start-date"
        max="2222-02-27"
        required="required"
        type="date"
      />
      <!---->
      <!---->
      <!---->
    </div>
  </div>
   
  <div
    aria-labelledby="__BVID__19___BV_label_"
    class="b-form-group form-group"
    id="__BVID__19_"
    role="group"
  >
    <label
      class="col-form-label pt-0"
      for="sprint-end-date"
      id="__BVID__19___BV_label_"
    >
      End date
    </label>
    <div
      class=""
    >
      <input
        aria-required="true"
        class="form-control"
        id="sprint-end-date"
        min="2222-02-22"
        required="required"
        type="date"
      />
      <!---->
      <!---->
      <!---->
    </div>
  </div>
   
  <button
    class="btn btn-primary"
    type="submit"
  >
    Save
  </button>
   
  <button
    class="btn btn-grey"
    type="button"
  >
    Cancel
  </button>
</form>
`;
